### ISSUE TITLE
android heart rate -- insufficient permission

### ISSUE THREAD

**Issue by noktilux** [dariosalvi78/cordova-plugin-health#230]
Heart rate retrieval works in iOS, but on android I get the following error callback:  "Application needs OAuth consent from the user".  I have also tested steps and there is no problem with that one on either platform.

Not sure if this is an issue with the plugin or if I am missing something.

**Comment by dariosalvi78** [comment]
You need to configure the Google APIs, have you gotten approval from Google?

**Comment by noktilux** [comment]
I have configured API Key and OAuth 2.0 Client ID in Google BUT, since I am still in development and just testing this feature right now, I have not sought approval.  When I use your plugin and seek authorization, I get a standard Google "authorization flow" asking me to select account and then allow permissions.  That works and I can get things like steps and weight, it's the heart rate which causes the issue.

I understand that this variable requires a higher level of permission and, ultimately, approval from Google, but I imagined a private test would be OK.  Otherwise, it makes it rather difficult to do the development work.  So I guess the question would be:  is there a way to make this work for test purposes -- restricted to particular account(s) or something like that?  Or maybe I've just missed an important configuration step somewhere?

**Comment by dariosalvi78** [comment]
When testing, you can access all data types, but it has to be with the same test user you have added on the Google cloud interface.
Make sure you have requested the right data types on the Google console. Depending on the version of the API you may need heart rate or body (haven't understood really why the difference). Let us know if you solve it!

**Comment by dariosalvi78** [comment]
When you create the [OAuth consent screen](https://support.google.com/cloud/answer/10311615) for the Google Fitness API, you need to specify the scopes, something like shown here: 
![image](https://user-images.githubusercontent.com/5367037/137593497-ad9c0892-39b7-4fbd-bf94-059a617656b3.png)

**Comment by noktilux** [comment]
Aha!  That may be it.  Going to try soon.  Thank you.

**Comment by noktilux** [comment]
one more thing as I continue to be frustrated...

for your consideration, a re-write of one of your javascript functions:

```javascript
Health.prototype.toFitActivity = function (act) {
	const acts={
		supported:{
			core_training:'strength_training',
			flexibility:'gymnastics',
			stairs:'stair_climbing',
			'wheelchair.walkpace':'wheelchair',
			'wheelchair.runpace':'wheelchair',
			'sleep.inBed':'sleep.awake'
		},
		unsupported:{
			archery:1,
			barre:1,
			bowling:1,
			fishing:1,
			functional_strength:1,
			hunting:1,
			lacrosse:1,
			mixed_metabolic_cardio:1,
			paddle_sports:1,
			play:1,
			preparation_and_recovery:1,
			snow_sports:1,
			softball:1,
			water_fitness:1,
			water_sports:1,
			wrestling:1
		}
	};
	var activity=acts.supported[act];
	if(!activity){
		activity=acts.unsupported[act];
		if(activity){
			activity='other'
		}else{
			activity=act
		}
	}
	return activity
};
```

**Comment by dariosalvi78** [comment]
thanks for the rewrite, mind sending a pull request?

**Comment by noktilux** [comment]
I wrote to Google and actually got a reply back!  Here is what they say:

> Testing phase projects do not require verification however, despite this setting, if you are referring to the "unverified app" screen you see when you authenticate, then it is working as intended.  Any projects including testing projects will get this message if they are trying to authenticate to a sensitive or restrict scope.

> In order to remove this, you have a few options such as verification, domain whitelisting, or setting it for internal use (requires Google Workspace).  Otherwise, you can continue to test as you would with the unverified app screen and 100 user cap intact.

My issue is that when I load the app to attached Android device (via "cordova run android" command), I do not get the "unverified app" screen they are talking about.

Any ideas?

**Comment by noktilux** [comment]
Still struggling with this, but I *may* be getting somewhere.  Dario, are you able to show me the Android part of your plugin.xml file?  In particular, I want to make sure I have the API version settings correct.  Also, can you confirm that the first argument passed to navigator.health.requestAuthorization should look like this (I only need read permission):

```javascript
[
  {
    read : ['activity', 'body', 'heart_rate', 'steps']
  }
]
```

It seems the problem is with that last part somehow because it does not ask me to grant permission for heart rate, but does for the other things.


![access_auth](https://user-images.githubusercontent.com/10899624/141645083-1250baa6-b8eb-4f5e-a051-2aa94281c742.png)

**Comment by noktilux** [comment]
Dario.

Later today, I am going to write a "minimal" (small as possible) application just to test the heart rate feature.  If it doesn't work I would then like to hire you to investigate by sending you the code, if that's OK?

**Comment by noktilux** [comment]
I have written a minimal test app which can be downloaded from the following link:

[https://cardiologica.org/7/health.zip](https://cardiologica.org/7/health.zip)

testing steps:
![test-app](https://user-images.githubusercontent.com/10899624/141693577-654f2649-2f48-4525-8caa-65cc9c9f0bee.png)

testing heart rate:
![heart-rate](https://user-images.githubusercontent.com/10899624/141693594-5868f6f9-a19b-4984-ae9a-1db9067f8552.png)

**Comment by dariosalvi78** [comment]
Hi, there is no "body" data type, you should not request authorization for data types that are not defined (well, the plugin should raise an exception).

In my app I first asked to get access to the heart rate data scope, but later discovered that the "body measurements" scope was enough. I don't know why, but my only explanation is that Google is still rolling out these new scopes and they're not 100% enforced yet.


If you ask about the configuration of my app, these are the SDK version I have right now in my gradle:

```
      defaultBuildToolsVersion="29.0.2" //String
      defaultMinSdkVersion=22 //Integer - Minimum requirement is Android 5.1
      defaultTargetSdkVersion=29 //Integer - We ALWAYS target the latest by default
      defaultCompileSdkVersion=29 //Integer - We ALWAYS compile with the latest by default
```

And the Play services:

```
implementation "com.google.android.gms:play-services-fitness:20.0.0"
implementation "com.google.android.gms:play-services-auth:19.0.0"
```

Permissions from my manifest, but I have several plugins installed:
```
    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
    <uses-permission android:name="android.permission.BLUETOOTH" />
    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
    <uses-feature android:name="android.hardware.location.gps" android:required="true" />
    <uses-permission android:name="android.permission.BODY_SENSORS" />
    <uses-permission android:name="android.permission.ACTIVITY_RECOGNITION" />
    <uses-permission android:name="android.gms.permission.ACTIVITY_RECOGNITION" />
    <queries>
        <package android:name="com.google.android.apps.fitness" />
    </queries>
    <uses-permission android:name="android.permission.ACCESS_NOTIFICATION_POLICY" />
    <uses-permission android:name="android.permission.RECEIVE_BOOT_COMPLETED" />
    <uses-permission android:name="android.permission.WAKE_LOCK" />
    <uses-permission android:name="android.permission.RECORD_AUDIO" />
    <uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
```

**Comment by dariosalvi78** [comment]
to test the plugin, the best idea is to create an empty cordova app, setup the Google cloud stuff then open up the console and run the plugin from there

**Comment by noktilux** [comment]
This all seems to boil down to the fact that at the oAuth stage, I am not asked/told about heart rate.  If I go to the Google API "playground" (https://developers.google.com/oauthplayground) and select the following:

![pg-1](https://user-images.githubusercontent.com/10899624/141695722-adcf1534-8fb5-4200-8229-e8bc6c36a1ed.png)

...the oauth pop-up then shows this:

![playground-hr](https://user-images.githubusercontent.com/10899624/141695733-248cf695-d598-4260-ae02-5fccd3b76563.png)

Using, you plugin, I get a similar Google interface BUT there is no reference to heart rate, even though I have listed it as one of the things I want.  That seems to be the issue.   Can you show me a screenshot of what comes up for you during the authorization stage with your app implementation?  I am going crazy here.


One more thing, when you say open up console and run the plugin from there, are you referencing running the "cordova run android  --target=ABCDE..." command or something else?

**Comment by dariosalvi78** [comment]
> This all seems to boil down to the fact that at the oAuth stage, I am not asked/told about heart rate.This all seems to boil down to the fact that at the oAuth stage, I am not asked/told about heart rate.

The plugin requests access to the data types as they are specified in the call to requestAuthorization(), see https://github.com/dariosalvi78/cordova-plugin-health/blob/688b1f4601826eb44ba39c5f858476a486c9e415/src/android/HealthPlugin.java#L382

In your case `heart_rate` is mapped to [DataType.TYPE_HEART_RATE_BPM](https://developers.google.com/android/reference/com/google/android/gms/fitness/data/DataType#TYPE_HEART_RATE_BPM), which, in terms of scope, should use the https://www.googleapis.com/auth/fitness.heart_rate.read scope, at least according to [this](https://developers.google.com/fit/datatypes/body#heart_rate).

Now what's happening here? I have no idea.

In my case, I too requested the heart_rate.read scope on Google cloud, but then the app requested the body scope instead and all worked. I even had to drop the heart_rate scope from Google cloud in order for the app to be approved, as it was useless.

Now, this was a few months ago though so maybe they have changed that.

In terms of the plugin, I don't think there's anything wrong: heart rate is treated like all the other data types. The issue should be either on your side or on Google's. Maybe you can try contacting them?

> Can you show me a screenshot of what comes up for you during the authorization stage with your app implementation?

I don't get the request for heart rate. As said, in my case it was enough with body measurements.

> when you say open up console and run the plugin from there

you can open Chrome DevTools and use the console to call, for example, navigator.health.requestAuthorization(..) no need to create any UI, just type in what you want to test

**Comment by noktilux** [comment]
> you can open Chrome DevTools and use the console to call, for example, navigator.health.requestAuthorization(..) no need to create any UI, just type in what you want to test

I just tried that and the exact same thing happened as in the minimal app I posted earlier.  As I asked earlier, can I hire you to try that and see what happens at your end?  I'm in a real bind here and will probably lose my contract soon if I can't get this working.

**Comment by dariosalvi78** [comment]
> the exact same thing happened as in the minimal app I posted earlier

I wasn't expecting anything different, it was just a suggestion about how to test plugins within cordova.

> will probably lose my contract soon if I can't get this working

I'm sorry to hear it, but unfortunately I'm very busy in this period. If I were you I'd try to find an Android developer who can setup a proof-of-concept Android-native app for reading heart rate. If it works and we find out that something is missing in the plugin I'll be happy to fix it.

Upwork could be good for finding short-term developers, I have used it successfully in the past.

**Comment by noktilux** [comment]
Dario.

I followed your advice and hired an experienced Android developer at Upwork.   It is now just over a week later and, trying this both independently and with my little "minimal" sample, he is also not able to get heart rate working.  

On top of that, I went through the effort of making yet another small test application that uses Google's REST API.  With that one, I can get it to work with heart rate!  Unfortunately, the method does not work in the context of an Android app.

Even though I have put in over 100 hours into this one problem, I am willing to accept I am missing some trivial little setup issue, but it sure is looking like some kind of context-specific bug.  In the end, whatever the cause, I feel that if someone like me (with 3 decades of coding experience) is stuck on this "issue", it's worth taking the time to figure out why.  It is very much in my interest to keep this plugin viable and I am willing to put even more effort into figuring out what is going wrong, contribute to some documentation, but I really need your help at this point OR the help of someone else who has managed to get heart rate working.

**Comment by dariosalvi78** [comment]
@noktilux  I am really sorry that you are encountering all these problems!
Let me do some little tests myself, at least we get to understand if the problem is on your setup or if it's on the plugin site.
I'll let you know.

**Comment by dariosalvi78** [comment]
I have tested the plugin and I can confirm that I can retrieve heart rate on Android.
This is what I have done:
- Create a new Cordova project
- Add Android platform (I am using the latest 10.1.1)
- Make sure you have the right SDK, build tools etc (use Android Studio for that)
- Compile the app to make sure it works
- Add the plugin
- Compile the app again to check that everything is OK
- Configure Google Cloud and make sure all keys / app IDs are correct under Credentials
- Configure the Oauth consent screen by selecting the appropriate scopes (I have selected them all just for testing) and by adding the correct test users
- Run the app on the device that has the same user as the test users
- Open chrome://inspect/
- Issue the commands from there to test:
  - Ask for permissions (I have issued read/write for steps and heart rate)
  - Query in the last day or so

Things to check:
* app ID
* test users
* right scopes
* right query (heart_rate is not compatible with queryAggregated)

Once a user has granted access to the scope, the only way to remove the authorization is to go to https://myaccount.google.com/permissions

I'll close the issue now, good luck!

**Comment by noktilux** [comment]
Dario, I have been going over my Google Cloud setup very carefully and looking again carefully at the screenshot of "sensitive scopes" you posted in one of the comments above, something is very odd:  you do NOT have heart rate!

![dario-scopes](https://user-images.githubusercontent.com/10899624/143647484-3ec49f5d-9bde-4e1c-9d3c-069a06fa2a59.png)

By contrast, this is what mine look like (at the moment -- I have tried taking everything in and out, one by one)

![sensitive-scopes](https://user-images.githubusercontent.com/10899624/143648227-2e5783da-cad6-4f7c-be36-4eb23b579384.png)

Is it possible, because my setup is rather recent, that Google have changed and broken something?  It would have to mean that your permissions are somehow "grandfathered", if that's the right expression here.   

When you ran your most recent test, was it with your existing project/account or did you create something new on Google Cloud?

**Comment by dariosalvi78** [comment]
in this test I have done today I have used an existing project, but went through the setup from the start, like starting anew. To be clear: I left the Oauth process in testing mode, it's not a project I use in production.

I would suggest, for the sake of testing, that you add all the available scopes, read and write. Once you have it working you can go back to Google Cloud and remove stuff that you don't need until you find the culprit. It's a bit of a painful process, but at least it will allow you to solve it hopefully.

**Comment by noktilux** [comment]
Dario.

After setting up a clean install of everything, including Ubuntu 20.04, on a new machine, I was able to get everything working.  Yay!  It was nothing to do with my code, your code or Google Cloud config, everything to do with older versions of software.

The computer I was using for development had installs/configurations of Cordova, Android, and Gradle that dated back to late 2019.  I have now updated all those -- a task that had included its own headaches -- and now have that working as well.

When I have a bit of time, I am going to put together a little Linux install guide to get this going with emphasis on the Cordova-related aspects.
